common/interpolation/src/spline_interpolation.cpp

```diff
@@ -236,13 +243,38 @@ std::vector<double> SplineInterpolation::getSplineInterpolatedValues(
       ++j;
     }
 
-    const double ds = query_key - base_keys_.at(j);
+    const double & ds = query_key - base_keys_.at(j);
     res.push_back(d.at(j) + (c.at(j) + (b.at(j) + a.at(j) * ds) * ds) * ds);
   }
 
   return res;
 }
 
+double SplineInterpolation::interpolatePoint(const double & query_key) const
+{
+  if (base_keys_.empty()) {
+    throw std::domain_error("The base keys are not available, please initialize the keys.");
+  }
+
+  const auto & a = multi_spline_coef_.a;
+  const auto & b = multi_spline_coef_.b;
+  const auto & c = multi_spline_coef_.c;
+  const auto & d = multi_spline_coef_.d;
+
+  // Binary search the minimum element in the base_key in such way that : query_key <= base_key[i]
+  auto const lower_bound_it = std::lower_bound(base_keys_.cbegin(), base_keys_.cend(), query_key);
+
+  // k1 is the index query_key <=base_keys[k1]
+  auto const & k1 = std::distance(base_keys_.cbegin(), lower_bound_it);
+  auto const & k0 = k1 == 0 ? k1 : k1 - 1;  // if upper bound coincides with the first element
+
+  // Get the interpolated value
+  const double & ds = query_key - base_keys_.at(k0);
+  auto const & query_value = d.at(k0) + (c.at(k0) + (b.at(k0) + a.at(k0) * ds) * ds) * ds;
+
+  return query_value;
+}
```

Comment:
This function's name should be the same as `getSplineInterpolatedValues`.

In addition, to remove duplicated implementation, this function should be like
```
std::vector<double> query_keys(query_key);
return getSplineInterpolatedValues(query_keys.front()).front();
```

What do you think.

Reply:
This breaks other codes. I tried. 
/mnt/WD/Codes/ROS2_AWF/src/universe/autoware.universe/common/interpolation/src/spline_interpolation_points_2d.cpp:131:69: error: request for member ‘at’ in ‘((const SplineInterpolationPoints2d*)this)->SplineInterpolationPoints2d::spline_y_.SplineInterpolation::getSplineInterpolatedValues(whole_s)’, which is of non-class type ‘double’
  131 |   const double y = spline_y_.getSplineInterpolatedValues({whole_s}).at(0);